File: 1erdepartamental/c++/20.merge_indirecto.cpp
// ============================================
// PROGRAMA: Merge Sort Indirecto
// DESCRIPCIÓN: Ordena índices con divide y conquista
// ============================================

#include <iostream>
#include <vector>
using namespace std;


/**
 * Merge sort indirecto: ordena índices, preserva original
 */
template<typename T>
class OrdenadorIndirectoMerge {
private:
    const vector<T>* __datosOriginales;
    vector<int> __indices;
    
    void fusionar(int izq, int medio, int der) {
        int n1 = medio - izq + 1;
        int n2 = der - medio;
        
        vector<int> izqArr(n1), derArr(n2);
        
        for (int i = 0; i < n1; i++)
            izqArr[i] = __indices[izq + i];
        for (int j = 0; j < n2; j++)
            derArr[j] = __indices[medio + 1 + j];
        
        int i = 0, j = 0, k = izq;
        
        while (i < n1 && j < n2) {
            // Compara valores originales usando índices
            if ((*__datosOriginales)[izqArr[i]] <= 
                (*__datosOriginales)[derArr[j]]) {
                __indices[k] = izqArr[i];
                i++;
            } else {
                __indices[k] = derArr[j];
                j++;
            }
            k++;
        }
        
        while (i < n1) __indices[k++] = izqArr[i++];
        while (j < n2) __indices[k++] = derArr[j++];
    }
    
    void mergeSortRecursivo(int izq, int der) {
        if (izq >= der) return;
        
        int medio = izq + (der - izq) / 2;
        
        mergeSortRecursivo(izq, medio);
        mergeSortRecursivo(medio + 1, der);
        fusionar(izq, medio, der);
    }

public:
    OrdenadorIndirectoMerge(const vector<T>& datos)
        : __datosOriginales(&datos) {
        for (size_t i = 0; i < datos.size(); i++) {
            __indices.push_back(i);
        }
    }
    
    void ordenar() {
        if (__indices.size() > 1) {
            mergeSortRecursivo(0, __indices.size() - 1);
        }
    }
    
    vector<int> getIndices() const { return __indices; }
    
    void mostrarOrdenado() const {
        cout << "Orden mediante índices: ";
        for (int idx : __indices) {
            cout << (*__datosOriginales)[idx] << " ";
        }
        cout << endl;
    }
    
    void mostrarOriginal() const {
        cout << "Array original: ";
        for (const T& val : *__datosOriginales) {
            cout << val << " ";
        }
        cout << endl;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    vector<int> numeros = {64, 34, 25, 12, 22, 11, 90};
    
    OrdenadorIndirectoMerge<int> ordenador(numeros);
    
    ordenador.mostrarOriginal();
    ordenador.ordenar();
    ordenador.mostrarOrdenado();
    ordenador.mostrarOriginal();
    
    return 0;
}

File: 1erdepartamental/python/20.merge_indirecto.py
# ============================================
# PROGRAMA: Merge Sort Indirecto
# DESCRIPCIÓN: Fusión ordenando índices
# ============================================

from abc import ABC, abstractmethod


class OrdenadorIndirecto(ABC):
    """Base para ordenamiento indirecto"""

    def __init__(self, datos):
        self.__datos = datos
        self.__indices = list(range(len(datos)))

    @property
    def datos_originales(self):
        return self.__datos

    @property
    def indices(self):
        return self.__indices.copy()

    @abstractmethod
    def _ordenar_indices_impl(self, idx):
        pass

    def ordenar(self):
        self.__indices = self._ordenar_indices_impl(self.__indices)
        return self.__indices

    def acceder_ordenado(self):
        return [self.__datos[i] for i in self.__indices]


class OrdenadorMergeSortIndirecto(OrdenadorIndirecto):
    """Merge Sort que ordena índices"""

    def _ordenar_indices_impl(self, idx):
        if len(idx) <= 1:
            return idx

        medio = len(idx) // 2
        izq = self._ordenar_indices_impl(idx[:medio])
        der = self._ordenar_indices_impl(idx[medio:])

        return self._fusionar_indices(izq, der)

    def _fusionar_indices(self, izq, der):
        """Fusiona comparando datos[índice]"""
        resultado = []
        i = j = 0

        while i < len(izq) and j < len(der):
            # Compara usando índices para acceder datos
            if (
                self._OrdenadorIndirecto__datos[izq[i]]
                <= self._OrdenadorIndirecto__datos[der[j]]
            ):
                resultado.append(izq[i])
                i += 1
            else:
                resultado.append(der[j])
                j += 1

        resultado.extend(izq[i:])
        resultado.extend(der[j:])

        return resultado


class Persona:
    def __init__(self, nombre, edad):
        self.__nombre = nombre
        self.__edad = edad

    @property
    def nombre(self):
        return self.__nombre

    @property
    def edad(self):
        return self.__edad

    def __le__(self, otro):
        return self.__edad <= otro.edad

    def __str__(self):
        return f"{self.__nombre}({self.__edad})"


class OrdenadorMergeSortIndirectoPersonas(OrdenadorMergeSortIndirecto):
    pass


# ============================================
# EJECUCIÓN
# ============================================
if __name__ == "__main__":
    datos = [64, 34, 25, 12, 22, 11, 90]
    o = OrdenadorMergeSortIndirecto(datos)
    o.ordenar()

    print(f"Datos originales: {datos}")
    print(f"Acceso ordenado: {o.acceder_ordenado()}")

    print()

    personas = [Persona("Juan", 25), Persona("Ana", 30), Persona("Pedro", 20)]
    op = OrdenadorMergeSortIndirectoPersonas(personas)
    op.ordenar()
    print(f"Por edad: {[str(p) for p in op.acceder_ordenado()]}")
